processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java

```diff
@@ -234,13 +235,17 @@ public Metadata getMetadata()
       // Virtual columns cannot depend on each other, so we don't need to check transitive dependencies.
       if (baseColumns.containsAll(virtualColumn.requiredColumns())) {
         preJoinVirtualColumns.add(virtualColumn);
+        // Since pre-join virtual columns can be computed using only base columns, we include them in the
+        // base column set.
+        baseColumns.add(virtualColumn.getOutputName());
```

Comment:
missing unit tests for this in `HashJoinSegmentStorageAdapterTest`

Reply:
This case is covered by `JoinFilterAnalyzerTest.test_filterPushDown_factToRegionToCountryLeftFilterOnChannelVirtualColumn`